Shopify/ruby-lsp
# Add code lens for running entire test class

### Motivation

The first step of #584 

Using [code lens](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeLens), identify test classes and (in combination https://github.com/Shopify/vscode-ruby-lsp/pull/506) allow users to run all of the tests in that class by clicking "Run" 
- See gif below for demo

The tests are run in the terminal with the command `bundle exec ruby -Itest  <test_path>` (or just `ruby` if there is no `Gemfile`)

### Implementation

- Create new request `CodeLens`
- Visit every class in the document and check if the name of the class ends with `Test`
  - If so, adds a new `Interface::CodeLens` to the results array
  - `Interface::CodeLens` contains the range, command information, and data (`type: "test"`)
  - The `command` is a `Interface::Command` that runs the VSCode command `rubyLsp.runTest` (defined here) with the terminal command passed in as an argument
 - The `command` will be visible to users when they are on a file with a test class (see gif demo below)
 - When "Run" is clicked, [`rubyLsp.runTest`](https://github.com/Shopify/vscode-ruby-lsp/pull/506/files#diff-25d66d74617fe2e23d7946bd6e3ba95640ab1b9bc8947445d604fc271c7c1f12R243) is called and the entire test class is run from the terminal

### Automated Tests

Added tests for the new functionality.

### Manual Tests

See manual test instructions in https://github.com/Shopify/vscode-ruby-lsp/pull/506.
